osquery/tables/system/posix/augeas.cpp

```diff
@@ -130,68 +79,118 @@ void matchAugeasPattern(augeas* aug,
+      path = path.substr(6);
+      // The caller is responsible for the matching memory.
+      free(file);
+    } else {
+      // The iterator is currently pointing to a folder so we extract the path
+      // from the node.
+      path = node.substr(6);
     }
+
+    results.emplace_back(
+        std::initializer_list<std::pair<const std::string, std::string>>{
+            {"node", node},
+            {"value", value == nullptr ? "" : value},
+            {"path", path},
+            {"label", label == nullptr ? "" : label}});
   }
 
   // aug_match() allocates the matches array and expects the caller to free it.
   free(matches);
 }
 
-QueryData genAugeas(QueryContext& context) {
-  augeas* aug = aug_init(
-      nullptr, FLAGS_augeas_lenses.c_str(), AUG_NO_ERR_CLOSE | AUG_ENABLE_SPAN);
+class AugeasHandle {
+ public:
+  augeas* aug{nullptr};
+  bool error{false};
+
+  void initialize() {
+    std::call_once(initialized, [this]() {
+      this->aug = aug_init(
+          nullptr, FLAGS_augeas_lenses.c_str(), AUG_NO_ERR_CLOSE | AUG_NO_LOAD);
+      // Handle initialization errors.
+      if (this->aug == nullptr) {
+        LOG(ERROR) << "An error has occurred while trying to initialize augeas";
+        error = true;
+      } else if (aug_error(this->aug) != AUG_NOERROR) {
+        error = true;
+        // Do not use aug_error_details() here since augeas is not fully
+        // initialized.
+        LOG(ERROR)
+            << "An error has occurred while trying to initialize augeas: "
+            << aug_error_message(this->aug);
+        aug_close(this->aug);
+      }
+    });
+  }
 
-  // Handle initialization errors.
-  if (aug == nullptr) {
-    VLOG(1) << "An error has occurred while trying to initialize augeas";
-    return {};
-  } else if (aug_error(aug) != AUG_NOERROR) {
-    // Do not use aug_error_details() here since augeas is not fully
-    // initialized.
-    VLOG(1) << "An error has occurred while trying to initialize augeas: "
-            << aug_error_message(aug);
+  ~AugeasHandle() {
     aug_close(aug);
+  }
+
+ private:
+  std::once_flag initialized;
+};
+
+static AugeasHandle kAugeasHandle;
+
+QueryData genAugeas(QueryContext& context) {
+  kAugeasHandle.initialize();
+
+  if (kAugeasHandle.error == true) {
     return {};
   }
 
+  augeas* aug = kAugeasHandle.aug;
+  aug_load(aug);
+
   QueryData results;
+  std::unordered_set<std::string> patterns;
+
+  if (context.hasConstraint("node", EQUALS)) {
+    auto nodes = context.constraints["node"].getAll(EQUALS);
+    patterns.insert(nodes.begin(), nodes.end());
+  }
+
   if (context.hasConstraint("path", EQUALS)) {
     // Allow requests via filesystem path.
-    // We will request the pattern match by path using an optional argument.
     auto paths = context.constraints["path"].getAll(EQUALS);
-    for (const auto& path : paths) {
-      matchAugeasPattern(aug, path, results, context, true);
+    std::ostringstream pattern;
+
+    for (auto path : paths) {
```

Comment:
Use `for (const auto& path : paths) {` unless you're moving the contents of paths, then don't `const` it.